File: data/src/main/java/home/oleg/placesnearme/repositories/VenueRepositoryImpl.java
package home.oleg.placesnearme.repositories;

import home.oleg.placesnearme.core_network.service.IFourSquareAPI;
import home.oleg.placesnearme.core_network.models.reposnses.ExploreResponse;
import home.oleg.placesnearme.core_network.models.reposnses.Response;
import home.oleg.placesnearme.core_network.models.reposnses.VenuesResponse;

import java.util.List;
import java.util.Map;

import home.oleg.placenearme.models.Section;
import home.oleg.placenearme.models.Venue;
import home.oleg.placenearme.repositories.VenueRepository;
import home.oleg.placenearme.repositories.VenueRequestParams;
import home.oleg.placesnearme.mapper.VenueMapper;
import io.reactivex.Single;

public class VenueRepositoryImpl implements VenueRepository {

    private final IFourSquareAPI api;
    private final QueryParamCreator queryParamCreator = new QueryParamCreator();

    public VenueRepositoryImpl(IFourSquareAPI api) {
        this.api = api;
    }

    @Override
    public Single<List<Venue>> getRecommendedBySection(Section section, VenueRequestParams filter) {
        Map<String, String> queryMap = queryParamCreator.create(section, filter);

        return api.explore(queryMap)
                .map(Response::getResponse)
                .map(ExploreResponse::getVenues)
                .map(VenueMapper::map);
    }

    @Override
    public Single<List<Venue>> search(String query, VenueRequestParams filter) {
        Map<String, String> queryMap = queryParamCreator.create(query, filter);
        return api.search(queryMap)
                .map(Response::getResponse)
                .map(VenuesResponse::getVenues)
                .map(VenueMapper::map);
    }

}

File: core-data/src/main/java/home/oleg/placesnearme/coredata/repositories/VenueRepositoryImpl.kt
package home.oleg.placesnearme.coredata.repositories

import home.oleg.placesnearme.coredata.mapper.VenueMapper
import home.oleg.placesnearme.coredomain.models.Section
import home.oleg.placesnearme.coredomain.models.Venue
import home.oleg.placesnearme.coredomain.repositories.VenueRepository
import home.oleg.placesnearme.coredomain.repositories.VenueRequestParams
import home.oleg.placesnearme.corenetwork.service.IFourSquareAPI
import io.reactivex.Single
import javax.inject.Inject

class VenueRepositoryImpl @Inject constructor(private val api: IFourSquareAPI) : VenueRepository {
    private val queryParamCreator = QueryParamCreator()

    override fun getRecommendedBySection(section: Section, requestParams: VenueRequestParams): Single<List<Venue>> {
        val queryMap = queryParamCreator.create(section, requestParams)

        return api.explore(queryMap)
                .map { it.response }
                .map { it.venues }
                .map { VenueMapper.map(it) }
    }

    override fun search(query: String, requestParams: VenueRequestParams): Single<List<Venue>> {
        val queryMap = queryParamCreator.create(query, requestParams)
        return api.search(queryMap)
                .map { it.response }
                .map { it.venues }
                .map { VenueMapper.map(it) }
    }

}
